incr. comp. ICE: try_mark_green() - Forcing the DepNode should have set its color

Note that this code doesn't compile regardless (`ast::Decl` contains `ast::Expr` by value, which contains an `ast::Decl`), but the ICE looks related to incremental compilation, so it might be reproducible under other circumstances.

[Code](https://github.com/oftlisp/oftb/tree/ice), [Travis Reproduction](https://travis-ci.org/oftlisp/oftb/jobs/354142303#L524)

```
$ cargo check
   Compiling oftb v0.1.0 (file:///home/nathan/Projects/oftlisp/oftb)
warning: unused import: `literal::Literal`
  --> src/flatanf/convert.rs:10:5
   |
10 | use literal::Literal;
   |     ^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

error[E0391]: unsupported cyclic reference between types/traits detected
  --> src/ast/mod.rs:62:1
   |
62 | pub enum Decl {
   | ^^^^^^^^^^^^^ cyclic reference
   |
note: the cycle begins when processing `ast::Decl`...
  --> src/ast/mod.rs:62:1
   |
62 | pub enum Decl {
   | ^^^^^^^^^^^^^
   = note: ...which then again requires processing `ast::Decl`, completing the cycle.

error: internal compiler error: /checkout/src/librustc/dep_graph/graph.rs:615: try_mark_green() - Forcing the DepNode should have set its color

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.24.1 (d3ae9a9e0 2018-02-27) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:504:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `oftb`.

To learn more, run the command again with --verbose.
```

Running this on `x86_64-pc-windows-msvc` gives me:
```
warning: unused import: `literal::Literal`
  --> src\flatanf\convert.rs:10:5
   |
10 | use literal::Literal;
   |     ^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

error[E0072]: recursive type `ast::Decl` has infinite size
  --> src\ast\mod.rs:62:1
   |
62 | pub enum Decl {
   | ^^^^^^^^^^^^^ recursive type has infinite size
63 |     Def(Symbol, Expr),
   |                 ----- recursive without indirection
64 |     Defn(Symbol, Vec<Symbol>, Vec<Expr>, Expr),
   |                                          ----- recursive without indirection
   |
   = help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `ast::Decl` representable

error[E0072]: recursive type `ast::Expr` has infinite size
   --> src\ast\mod.rs:112:1
    |
112 | pub enum Expr {
    | ^^^^^^^^^^^^^ recursive type has infinite size
113 |     Call(Box<Expr>, Vec<Expr>),
114 |     Decl(Decl),
    |          ----- recursive without indirection
    |
    = help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `ast::Expr` representable

error: aborting due to 2 previous errors
```
It seems like travis keeps around the incremental cache. Does this happen for clean builds too?

--- a/src/librustc/dep_graph/graph.rs
+++ b/src/librustc/dep_graph/graph.rs
@@ -648,8 +648,15 @@ impl DepGraph {
                                 return None
                             }
                             None => {
-                                bug!("try_mark_green() - Forcing the DepNode \
-                                      should have set its color")
+                                if !tcx.sess.has_errors() {
+                                    bug!("try_mark_green() - Forcing the DepNode \
+                                          should have set its color")
+                                } else {
+                                    // If the query we just forced has resulted
+                                    // in some kind of compilation error, we
+                                    // don't expect that the corresponding
+                                    // dep-node color has been updated.
+                                }
                             }
                         }
                     } else {

